File: scripts/generate-godot-art-assets.mjs
import { mkdir, writeFile } from "node:fs/promises";
import { dirname, join } from "node:path";

const root = "godot-client/assets";

function svg(width, height, body) {
  return `<svg xmlns="http://www.w3.org/2000/svg" width="${width}" height="${height}" viewBox="0 0 ${width} ${height}">
  <defs>
    <linearGradient id="gold" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0" stop-color="#fff0a8"/>
      <stop offset="1" stop-color="#d58a34"/>
    </linearGradient>
    <linearGradient id="sky" x1="0" y1="0" x2="0" y2="1">
      <stop offset="0" stop-color="#8fd8ff"/>
      <stop offset="0.48" stop-color="#bdeaff"/>
      <stop offset="1" stop-color="#77c56d"/>
    </linearGradient>
    <linearGradient id="steel" x1="0" y1="0" x2="1" y2="1">
      <stop offset="0" stop-color="#eff8ff"/>
      <stop offset="1" stop-color="#84a9c8"/>
    </linearGradient>
  </defs>
${body}
</svg>`;
}

function iconBase(body) {
  return svg(96, 96, `
  <rect x="6" y="6" width="84" height="84" rx="18" fill="#172437" stroke="#78b7ff" stroke-width="4"/>
  <ellipse cx="48" cy="74" rx="30" ry="8" fill="#07101b" opacity="0.5"/>
${body}`);
}

const files = {
  "backgrounds/ro_field_day.svg": svg(432, 768, `
  <rect width="432" height="768" fill="url(#sky)"/>
  <path d="M0 230 C90 170 170 210 238 158 C310 106 372 142 432 112 L432 768 L0 768 Z" fill="#6fb669"/>
  <path d="M0 290 C110 230 190 265 270 210 C340 165 390 184 432 160 L432 768 L0 768 Z" fill="#4f9e67"/>
  <path d="M-40 768 C80 590 168 510 232 400 C286 488 348 598 472 768 Z" fill="#bdad8c"/>
  <path d="M-20 768 C92 604 168 534 226 430 C278 510 336 610 452 768 Z" fill="#d8c6a2"/>
  <path d="M64 414 L368 414" stroke="#e8d5ae" stroke-width="8" opacity="0.7"/>
  <path d="M55 455 L378 455" stroke="#b29a77" stroke-width="4" opacity="0.55"/>
  <circle cx="68" cy="92" r="34" fill="#fff5b5" opacity="0.95"/>
  <g opacity="0.55" fill="#f4ffff">
    <ellipse cx="274" cy="84" rx="46" ry="14"/>
    <ellipse cx="318" cy="96" rx="58" ry="16"/>
    <ellipse cx="142" cy="122" rx="52" ry="15"/>
  </g>`),
  "sprites/player_adventurer.svg": svg(180, 220, `
  <ellipse cx="90" cy="202" rx="58" ry="12" fill="#07101b" opacity="0.45"/>
  <path d="M45 93 C22 128 22 168 50 196 L72 154 Z" fill="#426a94"/>
  <path d="M135 93 C158 128 158 168 130 196 L108 154 Z" fill="#2f557c"/>
  <path d="M58 72 C78 46 108 46 126 72 L138 156 C126 185 54 185 42 156 Z" fill="#70472f"/>
  <path d="M60 79 L120 79 L110 155 L70 155 Z" fill="url(#gold)"/>
  <rect x="62" y="108" width="56" height="18" fill="#7bd7ff" opacity="0.95"/>
  <circle cx="90" cy="47" r="28" fill="#f0d2a8"/>
  <path d="M58 45 C64 17 108 6 129 35 C119 26 93 32 79 40 C74 49 66 52 58 45 Z" fill="#1e2432"/>
  <path d="M132 92 L166 58" stroke="url(#steel)" stroke-width="8" stroke-linecap="round"/>
  <path d="M149 69 L174 45" stroke="#79d7ff" stroke-width="4" stroke-linecap="round"/>
  <path d="M48 154 L20 204" stroke="#ffe181" stroke-width="8" stroke-linecap="round"/>
  <path d="M16 205 L34 215" stroke="#7bd7ff" stroke-width="5" stroke-linecap="round"/>`),
  "sprites/lucky_chest.svg": svg(160, 130, `
  <ellipse cx="80" cy="116" rx="66" ry="10" fill="#07101b" opacity="0.42"/>
  <path d="M24 54 C34 20 126 20 136 54 L136 72 L24 72 Z" fill="#8a5738" stroke="#f3c96d" stroke-width="6"/>
  <rect x="22" y="60" width="116" height="50" rx="8" fill="#795033" stroke="#f3c96d" stroke-width="6"/>
  <rect x="68" y="24" width="24" height="86" fill="url(#gold)"/>
  <circle cx="80" cy="76" r="19" fill="#7bdcff" stroke="#eaffff" stroke-width="4"/>
  <path d="M18 48 L4 34 M142 48 L156 34 M80 15 L80 0" stroke="#fff0a8" stroke-width="6" stroke-linecap="round"/>`),
};

const icons = {
  pack: `<rect x="28" y="38" width="40" height="28" rx="4" fill="#8a5535" stroke="url(#gold)" stroke-width="5"/><rect x="43" y="25" width="10" height="43" fill="url(#gold)"/><circle cx="48" cy="52" r="11" fill="#6fd6ff"/>`,
  listing: `<path d="M24 42 L72 42 L64 25 L32 25 Z" fill="url(#gold)"/><rect x="28" y="42" width="40" height="24" fill="#27384d" stroke="#6fd6ff" stroke-width="4"/><path d="M34 69 L34 78 M62 69 L62 78" stroke="#6fd6ff" stroke-width="4"/>`,
  open: `<circle cx="48" cy="48" r="24" fill="#684394"/><path d="M48 16 L48 80 M16 48 L80 48 M25 25 L71 71 M71 25 L25 71" stroke="url(#gold)" stroke-width="5"/><circle cx="48" cy="48" r="11" fill="#6fd6ff"/>`,
  cooldown: `<path d="M48 22 A26 26 0 1 1 30 67" fill="none" stroke="#6fd6ff" stroke-width="7"/><path d="M48 48 L48 28 M48 48 L64 54" stroke="url(#gold)" stroke-width="6" stroke-linecap="round"/>`,
  market: `<rect x="22" y="40" width="52" height="26" rx="4" fill="#26384c" stroke="#6fd6ff" stroke-width="5"/><path d="M20 40 L76 40 L68 25 L28 25 Z" fill="url(#gold)"/><circle cx="36" cy="55" r="6" fill="#fff0a8"/><circle cx="60" cy="55" r="6" fill="#fff0a8"/>`,
  price: `<ellipse cx="42" cy="61" rx="20" ry="9" fill="#d58a34"/><ellipse cx="48" cy="49" rx="20" ry="9" fill="#f3c96d"/><ellipse cx="54" cy="37" rx="20" ry="9" fill="#fff0a8"/><text x="43" y="55" fill="#26384c" font-size="26" font-family="Arial" font-weight="700">G</text>`,
  lock: `<rect x="28" y="43" width="40" height="28" rx="4" fill="url(#gold)"/><path d="M34 45 V36 A14 14 0 0 1 62 36 V45" fill="none" stroke="#6fd6ff" stroke-width="7"/><circle cx="48" cy="56" r="5" fill="#26384c"/>`,
  force_open: `<path d="M48 16 L53 35 L72 24 L61 43 L82 48 L61 53 L72 72 L53 61 L48 82 L43 61 L24 72 L35 53 L14 48 L35 43 L24 24 L43 35 Z" fill="url(#gold)"/><rect x="34" y="34" width="28" height="28" fill="#8a5535" stroke="#6fd6ff" stroke-width="4"/>`,
  boss: `<circle cx="48" cy="52" r="25" fill="#82465e"/><circle cx="38" cy="47" r="5" fill="#ff7187"/><circle cx="58" cy="47" r="5" fill="#ff7187"/><path d="M28 34 L38 14 L43 38 M68 34 L58 14 L53 38" fill="url(#gold)"/>`,
  guild: `<path d="M48 16 L73 27 L66 65 L48 78 L30 65 L23 27 Z" fill="#38537a" stroke="#6fd6ff" stroke-width="5"/><path d="M48 24 V66 M34 40 H62" stroke="url(#gold)" stroke-width="7" stroke-linecap="round"/>`,
  ticket: `<path d="M23 35 H73 V61 H23 Z" fill="url(#gold)"/><circle cx="23" cy="48" r="8" fill="#172437"/><circle cx="73" cy="48" r="8" fill="#172437"/><path d="M48 37 V59" stroke="#172437" stroke-width="4" stroke-dasharray="4 4"/>`,
  season: `<circle cx="48" cy="38" r="20" fill="url(#gold)"/><path d="M36 56 L28 82 L47 66 L68 82 L60 56" fill="#6fd6ff"/><circle cx="48" cy="38" r="9" fill="#fff7bd"/>`,
  warrior: `<path d="M25 70 L70 25" stroke="url(#gold)" stroke-width="9" stroke-linecap="round"/><path d="M31 32 L66 67" stroke="#6fd6ff" stroke-width="6" stroke-linecap="round"/><path d="M66 18 L78 18 L78 30" fill="none" stroke="#eff8ff" stroke-width="5"/>`,
  ranger: `<path d="M58 20 C28 30 28 66 58 76" fill="none" stroke="url(#gold)" stroke-width="7"/><path d="M36 22 L66 74 M23 48 H75" stroke="#6fd6ff" stroke-width="5" stroke-linecap="round"/>`,
  mage: `<path d="M32 78 L63 18" stroke="url(#gold)" stroke-width="7" stroke-linecap="round"/><circle cx="65" cy="18" r="16" fill="#8b78ff"/><circle cx="65" cy="18" r="7" fill="#6fd6ff"/>`,
  cleric: `<circle cx="48" cy="48" r="25" fill="#356f68"/><rect x="43" y="25" width="10" height="46" fill="url(#gold)"/><rect x="25" y="43" width="46" height="10" fill="url(#gold)"/>`,
  skill: `<path d="M48 17 L58 39 L82 42 L64 58 L69 81 L48 69 L27 81 L32 58 L14 42 L38 39 Z" fill="url(#gold)" stroke="#6fd6ff" stroke-width="4"/>`,
};

for (const [name, body] of Object.entries(icons)) {
  files[`icons/${name}.svg`] = iconBase(body);
}

for (const [path, content] of Object.entries(files)) {
  const full = join(root, path);
  await mkdir(dirname(full), { recursive: true });
  await writeFile(full, content, "utf8");
}

console.log(`Generated ${Object.keys(files).length} Godot art assets.`);

File: scripts/generate-godot-art-assets.py
from pathlib import Path
from PIL import Image, ImageDraw
import math

ROOT = Path("godot-client/assets")


def ensure(path: str):
    full = ROOT / path
    full.parent.mkdir(parents=True, exist_ok=True)
    return full


def save(path: str, image: Image.Image):
    image.save(ensure(path))


def vertical_gradient(size, top, bottom):
    w, h = size
    image = Image.new("RGBA", size)
    draw = ImageDraw.Draw(image)
    for y in range(h):
        t = y / max(1, h - 1)
        color = tuple(int(top[i] * (1 - t) + bottom[i] * t) for i in range(4))
        draw.line((0, y, w, y), fill=color)
    return image


def draw_icon_bg(draw):
    draw.rounded_rectangle((6, 6, 90, 90), radius=18, fill=(23, 36, 55, 255), outline=(120, 183, 255, 255), width=4)
    draw.ellipse((18, 66, 78, 82), fill=(7, 16, 27, 115))


def new_icon():
    image = Image.new("RGBA", (96, 96), (0, 0, 0, 0))
    draw = ImageDraw.Draw(image)
    draw_icon_bg(draw)
    return image, draw


GOLD = (242, 205, 103, 255)
GOLD_DARK = (201, 128, 48, 255)
BLUE = (111, 214, 255, 255)
PANEL = (39, 56, 77, 255)


def background():
    image = vertical_gradient((432, 768), (143, 216, 255, 255), (102, 174, 95, 255))
    draw = ImageDraw.Draw(image)
    draw.polygon([(0, 250), (70, 198), (146, 225), (232, 170), (318, 132), (432, 110), (432, 768), (0, 768)], fill=(111, 182, 105, 255))
    draw.polygon([(0, 318), (90, 250), (172, 286), (266, 222), (354, 184), (432, 160), (432, 768), (0, 768)], fill=(79, 158, 103, 255))
    draw.polygon([(-40, 768), (80, 590), (168, 510), (232, 400), (286, 488), (348, 598), (472, 768)], fill=(216, 198, 162, 255))
    draw.polygon([(-12, 768), (100, 620), (174, 548), (228, 442), (276, 520), (336, 622), (444, 768)], fill=(188, 173, 140, 255))
    draw.ellipse((34, 58, 102, 126), fill=(255, 245, 181, 245))
    draw.ellipse((226, 68, 338, 104), fill=(244, 255, 255, 150))
    draw.ellipse((80, 110, 198, 146), fill=(244, 255, 255, 140))
    save("backgrounds/ro_field_day.png", image)


def player():
    image = Image.new("RGBA", (180, 220), (0, 0, 0, 0))
    draw = ImageDraw.Draw(image)
    draw.ellipse((32, 190, 148, 214), fill=(7, 16, 27, 110))
    draw.polygon([(45, 93), (22, 128), (22, 168), (50, 196), (72, 154)], fill=(66, 106, 148, 255))
    draw.polygon([(135, 93), (158, 128), (158, 168), (130, 196), (108, 154)], fill=(47, 85, 124, 255))
    draw.polygon([(58, 72), (78, 46), (108, 46), (126, 72), (138, 156), (126, 185), (54, 185), (42, 156)], fill=(112, 71, 47, 255))
    draw.polygon([(60, 79), (120, 79), (110, 155), (70, 155)], fill=GOLD)
    draw.rectangle((62, 108, 118, 126), fill=(123, 215, 255, 240))
    draw.ellipse((62, 19, 118, 75), fill=(240, 210, 168, 255))
    draw.pieslice((54, 8, 130, 72), 185, 360, fill=(30, 36, 50, 255))
    draw.line((132, 92, 166, 58), fill=(210, 232, 255, 255), width=8)
    draw.line((149, 69, 174, 45), fill=BLUE, width=4)
    draw.line((48, 154, 20, 204), fill=(255, 225, 129, 255), width=8)
    draw.line((16, 205, 34, 215), fill=BLUE, width=5)
    save("sprites/player_adventurer.png", image)


def chest():
    image = Image.new("RGBA", (160, 130), (0, 0, 0, 0))
    draw = ImageDraw.Draw(image)
    draw.ellipse((14, 106, 146, 126), fill=(7, 16, 27, 110))
    draw.rounded_rectangle((22, 54, 138, 112), radius=10, fill=(121, 80, 51, 255), outline=GOLD, width=6)
    draw.pieslice((22, 14, 138, 94), 180, 360, fill=(138, 87, 56, 255), outline=GOLD, width=6)
    draw.rectangle((68, 24, 92, 110), fill=GOLD)
    draw.ellipse((61, 57, 99, 95), fill=BLUE, outline=(235, 255, 255, 255), width=4)
    for start, end in [((18, 48), (4, 34)), ((142, 48), (156, 34)), ((80, 15), (80, 0))]:
        draw.line((start, end), fill=(255, 240, 168, 255), width=6)
    save("sprites/lucky_chest.png", image)


def icon_pack(name, painter):
    image, draw = new_icon()
    painter(draw)
    save(f"icons/{name}.png", image)


def generate_icons():
    icon_pack("pack", lambda d: [d.rectangle((28, 38, 68, 66), fill=(138, 85, 53, 255), outline=GOLD, width=5), d.rectangle((43, 25, 53, 68), fill=GOLD), d.ellipse((37, 41, 59, 63), fill=BLUE)])
    icon_pack("listing", lambda d: [d.polygon([(24, 42), (72, 42), (64, 25), (32, 25)], fill=GOLD), d.rectangle((28, 42, 68, 66), fill=PANEL, outline=BLUE, width=4), d.line((34, 69, 34, 78), fill=BLUE, width=4), d.line((62, 69, 62, 78), fill=BLUE, width=4)])
    icon_pack("open", lambda d: [d.ellipse((24, 24, 72, 72), fill=(104, 67, 148, 255)), *[d.line((48, 48, 48 + math.cos(i * math.tau / 8) * 34, 48 + math.sin(i * math.tau / 8) * 34), fill=GOLD, width=5) for i in range(8)], d.ellipse((37, 37, 59, 59), fill=BLUE)])
    icon_pack("cooldown", lambda d: [d.arc((22, 22, 74, 74), -70, 295, fill=BLUE, width=7), d.line((48, 48, 48, 28), fill=GOLD, width=6), d.line((48, 48, 64, 54), fill=GOLD, width=6)])
    icon_pack("market", lambda d: [d.polygon([(20, 40), (76, 40), (68, 25), (28, 25)], fill=GOLD), d.rounded_rectangle((22, 40, 74, 66), radius=4, fill=PANEL, outline=BLUE, width=5), d.ellipse((30, 49, 42, 61), fill=(255, 240, 168, 255)), d.ellipse((54, 49, 66, 61), fill=(255, 240, 168, 255))])
    icon_pack("price", lambda d: [d.ellipse((22, 52, 62, 70), fill=GOLD_DARK), d.ellipse((28, 40, 68, 58), fill=GOLD), d.ellipse((34, 28, 74, 46), fill=(255, 240, 168, 255)), d.rectangle((45, 37, 54, 60), fill=PANEL)])
    icon_pack("lock", lambda d: [d.rounded_rectangle((28, 43, 68, 71), radius=4, fill=GOLD), d.arc((34, 22, 62, 50), 180, 360, fill=BLUE, width=7), d.ellipse((43, 51, 53, 61), fill=PANEL)])
    icon_pack("force_open", lambda d: [d.regular_polygon((48, 48, 34), 8, fill=GOLD), d.rectangle((34, 34, 62, 62), fill=(138, 85, 53, 255), outline=BLUE, width=4)])
    icon_pack("boss", lambda d: [d.ellipse((23, 27, 73, 77), fill=(130, 70, 94, 255)), d.ellipse((33, 42, 43, 52), fill=(255, 113, 135, 255)), d.ellipse((53, 42, 63, 52), fill=(255, 113, 135, 255)), d.polygon([(28, 34), (38, 14), (43, 38)], fill=GOLD), d.polygon([(68, 34), (58, 14), (53, 38)], fill=GOLD)])
    icon_pack("guild", lambda d: [d.polygon([(48, 16), (73, 27), (66, 65), (48, 78), (30, 65), (23, 27)], fill=(56, 83, 122, 255), outline=BLUE), d.line((48, 24, 48, 66), fill=GOLD, width=7), d.line((34, 40, 62, 40), fill=GOLD, width=7)])
    icon_pack("ticket", lambda d: [d.rectangle((23, 35, 73, 61), fill=GOLD), d.ellipse((15, 40, 31, 56), fill=(23, 36, 55, 255)), d.ellipse((65, 40, 81, 56), fill=(23, 36, 55, 255)), d.line((48, 37, 48, 59), fill=(23, 36, 55, 255), width=4)])
    icon_pack("season", lambda d: [d.ellipse((28, 18, 68, 58), fill=GOLD), d.polygon([(36, 56), (28, 82), (47, 66)], fill=BLUE), d.polygon([(60, 56), (68, 82), (49, 66)], fill=(122, 166, 255, 255)), d.ellipse((39, 29, 57, 47), fill=(255, 247, 189, 255))])
    icon_pack("warrior", lambda d: [d.line((25, 70, 70, 25), fill=GOLD, width=9), d.line((31, 32, 66, 67), fill=BLUE, width=6)])
    icon_pack("ranger", lambda d: [d.arc((24, 18, 76, 78), 105, 255, fill=GOLD, width=7), d.line((36, 22, 66, 74), fill=BLUE, width=5), d.line((23, 48, 75, 48), fill=BLUE, width=5)])
    icon_pack("mage", lambda d: [d.line((32, 78, 63, 18), fill=GOLD, width=7), d.ellipse((49, 2, 81, 34), fill=(139, 120, 255, 255)), d.ellipse((58, 11, 72, 25), fill=BLUE)])
    icon_pack("cleric", lambda d: [d.ellipse((23, 23, 73, 73), fill=(53, 111, 104, 255)), d.rectangle((43, 25, 53, 71), fill=GOLD), d.rectangle((25, 43, 71, 53), fill=GOLD)])
    icon_pack("skill", lambda d: [d.regular_polygon((48, 48, 32), 5, rotation=-math.pi / 2, fill=GOLD, outline=BLUE)])


background()
player()
chest()
generate_icons()
print("Generated 20 Godot PNG art assets.")
